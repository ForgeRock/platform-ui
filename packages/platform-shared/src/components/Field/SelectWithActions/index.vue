<!-- Copyright (c) 2021 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <FrSelect
    v-bind="$attrs"
    v-on="$listeners">
    <template v-slot:beforeList>
      <div class="d-flex align-items-center justify-content-between pl-3 pr-2 py-2">
        <h5 class="mb-0">
          {{ addRowText }}
        </h5>
        <!--
          Triggered on click, indicates user would like to add an item to the select list
          @event add-item-clicked
        -->
        <BButton
          @click="$emit('add-item-clicked')"
          :aria-label="addLabel"
          :title="addLabel"
          size="sm"
          variant="light">
          <i
            aria-hidden="true"
            class="material-icons-outlined">
            add
          </i>
        </BButton>
      </div>
    </template>
    <template v-slot:option="{ option }">
      <div class="d-flex align-items-center justify-content-between w-100 pr-2">
        <div
          :title="option.text"
          class="text-truncate py-2 pl-3 pr-2">
          {{ option.text }}
        </div>
        <!--
          Triggered on click, indicates user would like to view or edit this select list item
          @event edit-item-clicked
          @property {string} the value (id) of the item to view/edit
        -->
        <BButton
          @click.stop="$emit('edit-item-clicked', option.value)"
          v-if="option.value !== '[Empty]'"
          :aria-label="editLabel"
          :title="editLabel"
          size="sm"
          variant="light">
          <i
            aria-hidden="true"
            class="material-icons-outlined">
            edit
          </i>
        </BButton>
      </div>
    </template>
  </FrSelect>
</template>

<script>
import { BButton } from 'bootstrap-vue';
import Select from '@forgerock/platform-shared/src/components/Field/Select';

/**
 *  Select input with actions for adding and editing items which fire events.
 *  Allows selection of one element in a dropdown.
 */
export default {
  name: 'SelectWithActions',
  components: {
    BButton,
    FrSelect: Select,
  },
  props: {
    addRowText: {
      type: String,
      default() {
        return this.$t('common.add');
      },
      required: false,
    },
    addLabel: {
      type: String,
      default() {
        return this.$t('common.add');
      },
      required: false,
    },
    editLabel: {
      type: String,
      default() {
        return this.$t('common.edit');
      },
      required: false,
    },
  },
};
</script>
