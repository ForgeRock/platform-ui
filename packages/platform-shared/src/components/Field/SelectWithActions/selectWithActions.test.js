/**
 * Copyright (c) 2021 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import BootstrapVue from 'bootstrap-vue';
import { createLocalVue, shallowMount, mount } from '@vue/test-utils';
import SelectWithActions from './index';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

const defaultProps = {
  addRowText: 'addRow?',
  addLabel: 'addButton?',
  editLabel: 'editButton?',
  id: '',
  errorMessages: [],
  name: '',
  description: '',
  isHtml: false,
  label: '',
  options: [
    { text: 'a', value: 'ID_a' },
    { text: 'b', value: 'ID_b' },
  ],
};

describe('SelectWithActions input', () => {
  it('loads the component', () => {
    const wrapper = shallowMount(SelectWithActions, {
      mocks: {
        $t: () => {},
      },
      propsData: {
        ...defaultProps,
      },
    });
    expect(wrapper.name()).toBe('SelectWithActions');
  });

  it('emits add and edit events', () => {
    const wrapper = mount(SelectWithActions, {
      localVue,
      mocks: {
        $t: () => {},
      },
      propsData: {
        ...defaultProps,
      },
    });

    wrapper.find('button[aria-label="addButton?"').trigger('click');

    expect(wrapper.emitted()['add-item-clicked']).toBeTruthy();
    expect(wrapper.emitted()['add-item-clicked'].length).toBe(1);

    wrapper.findAll('button[aria-label="editButton?"').at(1).trigger('click');

    expect(wrapper.emitted()['edit-item-clicked']).toBeTruthy();
    expect(wrapper.emitted()['edit-item-clicked'].length).toBe(1);
    expect(wrapper.emitted()['edit-item-clicked'][0][0]).toEqual('ID_b');
  });

  it('sets default props', () => {
    const wrapper = mount(SelectWithActions, {
      localVue,
      mocks: {
        $t: (string) => {
          const translations = {
            'common.add': 'add',
            'common.edit': 'edit',
          };
          return translations[string];
        },
      },
      propsData: {
        id: '',
        errorMessages: [],
        name: '',
        description: '',
        isHtml: false,
        label: '',
        options: [],
      },
    });

    expect(wrapper.vm.addRowText).toBe('add');
    expect(wrapper.vm.addLabel).toBe('add');
    expect(wrapper.vm.editLabel).toBe('edit');
  });
});
