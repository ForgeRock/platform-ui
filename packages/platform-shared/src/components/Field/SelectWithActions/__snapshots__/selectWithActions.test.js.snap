// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectWithActions input emits add and edit events: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) Vue 3's render function API has changed. You can opt-in to the new API with:

  configureCompat({ RENDER_FUNCTION: false })

  (This can also be done per-component via the "compatConfig" option.)
  Details: https://v3-migration.vuejs.org/breaking-changes/render-function-api.html",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) vm.$listeners has been removed. In Vue 3, parent v-on listeners are included in vm.$attrs and it is no longer necessary to separately use v-on="$listeners" if you are already using v-bind="$attrs". (Note: the Vue 3 behavior only applies if this compat config is disabled)
  Details: https://v3-migration.vuejs.org/breaking-changes/listeners-removed.html",
    "
",
    " at <SelectWithActions",
    "addRowText="addRow?"",
    "addLabel="addButton?"",
    "editLabel="editButton?"",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation WATCH_ARRAY) (2)",
    "
",
    " at <SelectInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <SelectWithActions",
    "addRowText="addRow?"",
    "addLabel="addButton?"",
    "editLabel="editButton?"",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation WATCH_ARRAY) (3)",
    "
",
    " at <VueMultiselect",
    "id="floatingLabelInput4"",
    "ref="vms"",
    "value=""",
    " ...",
    ">",
    "
",
    " at <InputLayout",
    "id="floatingLabelInput4"",
    "name=""",
    "description=""",
    " ...",
    ">",
    "
",
    " at <SelectInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <SelectWithActions",
    "addRowText="addRow?"",
    "addLabel="addButton?"",
    "editLabel="editButton?"",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "appear" with v-bind value \`false\` will render appear="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <Transition",
    "name="multiselect__loading"",
    ">",
    "
",
    " at <VueMultiselect",
    "id="floatingLabelInput4"",
    "ref="vms"",
    "value=""",
    " ...",
    ">",
    "
",
    " at <InputLayout",
    "id="floatingLabelInput4"",
    "name=""",
    "description=""",
    " ...",
    ">",
    "
",
    " at <SelectInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <SelectWithActions",
    "addRowText="addRow?"",
    "addLabel="addButton?"",
    "editLabel="editButton?"",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "persisted" with v-bind value \`false\` will render persisted="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <Transition",
    "name="multiselect__loading"",
    ">",
    "
",
    " at <VueMultiselect",
    "id="floatingLabelInput4"",
    "ref="vms"",
    "value=""",
    " ...",
    ">",
    "
",
    " at <InputLayout",
    "id="floatingLabelInput4"",
    "name=""",
    "description=""",
    " ...",
    ">",
    "
",
    " at <SelectInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <SelectWithActions",
    "addRowText="addRow?"",
    "addLabel="addButton?"",
    "editLabel="editButton?"",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_ENUMERATED_COERCION) Enumerated attribute "spellcheck" with v-bind value \`false\` will render the value as-is instead of coercing the value to "true" in Vue 3. Always use explicit "true" or "false" values for enumerated attributes. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_ENUMERATED_COERCION: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <VueMultiselect",
    "id="floatingLabelInput4"",
    "ref="vms"",
    "value=""",
    " ...",
    ">",
    "
",
    " at <InputLayout",
    "id="floatingLabelInput4"",
    "name=""",
    "description=""",
    " ...",
    ">",
    "
",
    " at <SelectInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <SelectWithActions",
    "addRowText="addRow?"",
    "addLabel="addButton?"",
    "editLabel="editButton?"",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (2)",
    "
",
    " at <Transition",
    "name="multiselect"",
    "persisted=""",
    ">",
    "
",
    " at <VueMultiselect",
    "id="floatingLabelInput4"",
    "ref="vms"",
    "value=""",
    " ...",
    ">",
    "
",
    " at <InputLayout",
    "id="floatingLabelInput4"",
    "name=""",
    "description=""",
    " ...",
    ">",
    "
",
    " at <SelectInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <SelectWithActions",
    "addRowText="addRow?"",
    "addLabel="addButton?"",
    "editLabel="editButton?"",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) Functional component <BButton> should be defined as a plain function in Vue 3. The "functional" option has been removed. NOTE: Before migrating to use plain functions for functional components, first make sure that all async components usage have been migrated and its compat behavior has been disabled.
  Details: https://v3-migration.vuejs.org/breaking-changes/functional-components.html",
    "
",
    " at <Transition",
    "name="multiselect"",
    "persisted=""",
    ">",
    "
",
    " at <VueMultiselect",
    "id="floatingLabelInput4"",
    "ref="vms"",
    "value=""",
    " ...",
    ">",
    "
",
    " at <InputLayout",
    "id="floatingLabelInput4"",
    "name=""",
    "description=""",
    " ...",
    ">",
    "
",
    " at <SelectInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <SelectWithActions",
    "addRowText="addRow?"",
    "addLabel="addButton?"",
    "editLabel="editButton?"",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (2)",
    "
",
    " at <BButton",
    "onClick=fn",
    "aria-label="addButton?"",
    "title="addButton?"",
    " ...",
    ">",
    "
",
    " at <Transition",
    "name="multiselect"",
    "persisted=""",
    ">",
    "
",
    " at <VueMultiselect",
    "id="floatingLabelInput4"",
    "ref="vms"",
    "value=""",
    " ...",
    ">",
    "
",
    " at <InputLayout",
    "id="floatingLabelInput4"",
    "name=""",
    "description=""",
    " ...",
    ">",
    "
",
    " at <SelectInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <SelectWithActions",
    "addRowText="addRow?"",
    "addLabel="addButton?"",
    "editLabel="editButton?"",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;
